### DISCUSSION TITLE
Rotating any geometry with mouse

### DISCUSSION THREAD

**Discussion by fcdalgic** [pelicanmapping/osgearth#2657] [Support Q&A] [answered]
Hello, I am trying to rotate a selected geometry with mouse events. Since geometry could be a Polygon, Polyline, Rectangle, Circle, etc, first I encapsulated them with MatrixTranform before adding them to the corresponding layer. After I picked the geometry from that layer I could successfully drag and move the object but could not be able to rotate them via this MatrixTransform.  (Two different pickers for each action)

I even tried to encapsulate with PositionAttiudeTransform and do the rotation, as well as encapsulate with GeoTransform. Also, tried to perform that rotation with two different matrix calculations but I could not figure out why it was not rotating and instead jumping in horizontal directions. 

I would appreciate it if you could show me what I am doing wrong. Thanks in advance.

#### Attempt-1

	bool RotateGeometry(double directionAngle)
	{

		auto rotationMt = GetMatrixTransform();

		if (!rotationMt)
			return false;

		osg::Quat rotation(-directionAngle, osg::Z_AXIS);
		osg::Matrixd currentMatrix = rotationMt->getMatrix();
		osg::Vec3d pivotPoint = currentMatrix.getTrans(); // Rotation pivot

		// Apply the rotation around the pivot point
		osg::Matrixd T_inv = osg::Matrix::translate(-pivotPoint);
		osg::Matrixd R = osg::Matrix::rotate(rotation);
		osg::Matrixd T = osg::Matrix::translate(pivotPoint);

		osg::Matrixd newMatrix = T * R * T_inv * currentMatrix;
		rotationMt->setMatrix(newMatrix);

		return true;
	}

#### Attempt-2

		bool RotateGeometry(double directionAngle)
		{

			auto rotationMt = GetMatrixTransform();

			if (!rotationMt)
				return false;

			osg::Quat rotation(-directionAngle, osg::Z_AXIS);
			osg::Matrixd currentMatrix = rotationMt->getMatrix();
			osg::Matrixd R = osg::Matrix::rotate(rotation);

			osg::Matrixd newMatrix = R * currentMatrix;
			rotationMt->setMatrix(newMatrix);

			return true;
		}

**Comment by gwaldron** [comment]
Hard to say without a runnable example. Could be matrix multiplication order but I'm sure you have already fiddled with that.

The GeoTransform + PositionAttitudeTransform approach is demonstrated in `osgearth_transform` here:
https://github.com/gwaldron/osgearth/blob/master/src/applications/osgearth_transform/osgearth_transform.cpp#L77

The GeoTransform will establish a local tangent place (ENU) to the earth at its location, and then any rotations under it are relative to that coordinate plane.

**Reply by fcdalgic** [reply]
I think I partially figured that.
For Circle and Rectangle geometries (RectangleNode, CircleNode) they are already derived from LocalGeometryNode/GeoPositionNode no additional transformation matrix is needed. I could directly use setLocalRotation and setPosition. In fact, if we encapsulate them with another transformation matrix, the local transformation is applied twice and the result was funny.

On the other hand, I am still looking for a solution on how to rotate FeatureNode? I normally construct Polygon or Polyline geometries with that node type where vertices are fed by WGS-84 latitude and longitude.  I am thinking about converting each point to the local coordinates and encapsulating FeatureNode with GeoPositionNode's PositionAttitudeTransform. But I am worried about the computation costs, do you have any suggestions?

Thanks in advance,

**Reply by gwaldron** [reply]
FeatureNode is meant to represent static Feature data which is expressed in absolute map coordinates. The concept of rotation does not really apply there...what does it mean to rotate a GIS feature? What's the reference frame? I'd suggest that you are better off just using a simple GeoTransform + PAT combo (or LocalGeometryNode if you want to use the styling).

**Reply by fcdalgic** [reply]
I'm sorry. I did not know the purpose of the FeatureNode. I have just followed the examples to draw polygons on a map. What I really want is to draw a polygon geometry on the map and be able to apply the transformation. Yes, I need to use styling, is it better to write a PolygonNode class derived from LocalGeometryNode like the CircleNode? 

Thanks for the help.

**Reply by gwaldron** [reply, marked as answer]
Yes, that is a better approach. Good luck :)
